src/Tokenizer.php

```diff
@@ -0,0 +1,997 @@
+        if (preg_match('/^(' . $this->regexBoundaries . ')/', $string, $matches)) {
+            return new Token(Token::TOKEN_TYPE_BOUNDARY, $matches[1]);
+        }
+
+        // A reserved word cannot be preceded by a '.'
+        // this makes it so in "mytable.from", "from" is not considered a reserved word
+        if (! $previous || $previous->value() !== '.') {
+            $upper = strtoupper($string);
+            // Top Level Reserved Word
+            if (preg_match(
+                '/^(' . $this->regexReservedToplevel . ')($|\s|' . $this->regexBoundaries . ')/',
+                $upper,
+                $matches
+            )) {
+                return new Token(
+                    Token::TOKEN_TYPE_RESERVED_TOPLEVEL,
+                    substr($string, 0, strlen($matches[1]))
+                );
+            }
+
+            // Newline Reserved Word
+            if (preg_match(
+                '/^(' . $this->regexReservedNewline . ')($|\s|' . $this->regexBoundaries . ')/',
+                $upper,
+                $matches
+            )) {
+                return new Token(
+                    Token::TOKEN_TYPE_RESERVED_NEWLINE,
+                    substr($string, 0, strlen($matches[1]))
+                );
+            }
+
+            // Other Reserved Word
+            if (preg_match(
+                '/^(' . $this->regexReserved . ')($|\s|' . $this->regexBoundaries . ')/',
+                $upper,
+                $matches
+            )) {
+                return new Token(
+                    Token::TOKEN_TYPE_RESERVED,
+                    substr($string, 0, strlen($matches[1]))
+                );
+            }
+        }
+
+        // A function must be suceeded by '('
+        // this makes it so "count(" is considered a function, but "count" alone is not
+        $upper = strtoupper($string);
+        // function
+        if (preg_match('/^(' . $this->regexFunction . '[(]|\s|[)])/', $upper, $matches)) {
+            return new Token(
+                Token::TOKEN_TYPE_RESERVED,
+                substr($string, 0, strlen($matches[1])-1)
+            );
+        }
+
+        // Non reserved word
+        preg_match('/^(.*?)($|\s|["\'`]|' . $this->regexBoundaries . ')/', $string, $matches);
+
+        return new Token(Token::TOKEN_TYPE_WORD, $matches[1]);
+    }
+
+    /**
+     * Helper function for building regular expressions for reserved words and boundary characters
+     *
+     * @param string[] $strings The strings to be quoted
+     *
+     * @return string[] The quoted strings
+     */
+    private function quoteRegex(array $strings) : array
+    {
+        return array_map(static function (string $string) : string {
+            return preg_quote($string, '/');
+        }, $strings);
+    }
+
+    /**
+     * Get stats about the token cache
+     *
+     * @return mixed[] An array containing the keys 'hits', 'misses', 'entries', and 'size' in bytes
+     */
+    public function getCacheStats() : array
```

Comment:
do we really need this method?

Reply:
It's needed for `tests/performance.php`,  and unit tests, but apart from that, not really, now. Doesn't really hurt though, does it?